Josh/solutions_2021/day_7/solution.py

```diff
@@ -0,0 +1,69 @@
+from math import ceil, floor
+from os.path import dirname, join
+from typing import Any, Dict, List, Tuple
+
+from solutions_2021 import get_filename
+
+CURRENT_DIR = dirname(__file__)
+
+
+def parse_input(file_name: str) -> List[int]:
+    with open(join(CURRENT_DIR, file_name), "r") as fp:
+        contents = fp.read().split(",")
+    return [int(v) for v in contents]
+
+
+def get_minimum_entry(v: Dict[int, int]) -> Tuple[int, int]:
+    min_value = min(v.values())
+    for key in v.keys():
+        if v[key] == min_value:
+            return key, v[key]
+    return -1, -1
+
+
+def part_one(file_name: str) -> Tuple[int, int]:
+    initial_positions = parse_input(file_name)
+    fuel_usages = {}
+    min_pos, max_pos = min(initial_positions), max(initial_positions)
```

Comment:
would be quicker if you sorted `initial_positions` once and then did
```python
min_pos, max_pos = initial_positions[0], initial_positions[-1]
```
?

Reply:
idk man my current implementation is O(DM) where D is the difference between the min and max positions and M is the number os initial positions. If I sorted it, then the complexity would still be limited by my awful brute force loop

Reply:
ah right so my suggestion's performance improvement is basically negligible.

One alternative approach I saw was to start in the middle (mean or median) then
* calculate total fuel usage
* move position 1 to the right or left
* calculate total fuel usage again, which should hopefully be lower
* if the total fuel usage increases, then you've reached the minimum

kind of like gradient descent

Reply:
Sounds interesting, I was going to try and find a closed form solution which would probably look like gradient descent after applying the method the guy used for finding out the mean was a good solution